File: php/src/VedaChangeStream.php
<?php

declare(strict_types=1);

namespace VedaDB;

/**
 * Change stream for watching database changes (CDC).
 */
class VedaChangeStream
{
    private VedaClient $client;
    private ?string $table;
    private ?string $operation; // INSERT, UPDATE, DELETE, or null for all
    private float $timeoutMs;
    private bool $running = false;

    /** @var callable|null */
    private $handler = null;

    /** @var list<array<string, mixed>> */
    private array $buffer = [];

    private int $maxBufferSize = 1000;

    public function __construct(
        VedaClient $client,
        ?string $table = null,
        ?string $operation = null,
        float $timeoutMs = 30000.0,
    ) {
        $this->client    = $client;
        $this->table     = $table;
        $this->operation = $operation;
        $this->timeoutMs = $timeoutMs;
    }

    /**
     * Start watching for changes.
     *
     * @param callable(array<string, mixed>): void $handler
     */
    public function watch(callable $handler, ?float $timeoutMs = null): void
    {
        $this->handler = $handler;
        $this->running = true;

        $timeout = $timeoutMs ?? $this->timeoutMs;
        $endTime = microtime(true) + ($timeout / 1000);

        // Build WATCH command
        $cmd = 'WATCH';
        if ($this->table !== null) {
            $cmd .= " TABLE '{$this->table}'";
        }
        if ($this->operation !== null) {
            $cmd .= " FOR {$this->operation}";
        }

        $this->client->query($cmd);

        while ($this->running && microtime(true) < $endTime) {
            try {
                $result = $this->client->query('WATCH POLL');
                $change = $result->first();

                if ($change !== null) {
                    $this->buffer[] = $change;
                    if (count($this->buffer) > $this->maxBufferSize) {
                        array_shift($this->buffer);
                    }
                    ($handler)($change);
                }
            } catch (TimeoutException $e) {
                continue;
            } catch (\Throwable $e) {
                $this->running = false;
                throw $e;
            }
        }
    }

    /**
     * Poll for a single change event.
     *
     * @return array<string, mixed>|null
     */
    public function poll(?float $timeoutMs = null): ?array
    {
        $timeout = $timeoutMs ?? $this->timeoutMs;
        $endTime = microtime(true) + ($timeout / 1000);

        $cmd = 'WATCH';
        if ($this->table !== null) {
            $cmd .= " TABLE '{$this->table}'";
        }
        if ($this->operation !== null) {
            $cmd .= " FOR {$this->operation}";
        }

        while (microtime(true) < $endTime) {
            try {
                $result = $this->client->query($cmd . ' POLL');
                $change = $result->first();
                if ($change !== null) {
                    $this->buffer[] = $change;
                    return $change;
                }

                // Small backoff between polls
                usleep(100000); // 100ms
            } catch (TimeoutException $e) {
                return null;
            }
        }

        return null;
    }

    /**
     * Stop the change stream.
     */
    public function stop(): void
    {
        $this->running = false;
        try {
            $this->client->query('WATCH STOP');
        } catch (\Throwable) {
            // Ignore
        }
    }

    /**
     * Get buffered changes.
     *
     * @return list<array<string, mixed>>
     */
    public function getBuffer(): array
    {
        return $this->buffer;
    }

    /**
     * Get the last change.
     *
     * @return array<string, mixed>|null
     */
    public function last(): ?array
    {
        $count = count($this->buffer);
        return $count > 0 ? $this->buffer[$count - 1] : null;
    }

    /**
     * Resume from a resume token.
     */
    public function resume(string $token): void
    {
        $cmd = 'WATCH';
        if ($this->table !== null) {
            $cmd .= " TABLE '{$this->table}'";
        }
        $cmd .= " RESUME '{$token}'";
        $this->client->query($cmd);
    }

    /**
     * Check if the stream is running.
     */
    public function isRunning(): bool
    {
        return $this->running;
    }

    /**
     * Get the table being watched.
     */
    public function getTable(): ?string
    {
        return $this->table;
    }

    /**
     * Get the operation filter.
     */
    public function getOperation(): ?string
    {
        return $this->operation;
    }
}

File: dotnet/VedaChangeStream.cs
using System;
namespace VedaDB;

/// <summary>
/// Type of change event.
/// </summary>
public enum ChangeEventType
{
    /// <summary>A row was inserted.</summary>
    Insert,
    /// <summary>A row was updated.</summary>
    Update,
    /// <summary>A row was deleted.</summary>
    Delete,
    /// <summary>Table schema changed.</summary>
    Schema,
    /// <summary>Unknown change type.</summary>
    Unknown
}

/// <summary>
/// A change event from a watched table.
/// </summary>
public class VedaChangeEvent
{
    /// <summary>
    /// Type of change.
    /// </summary>
    public ChangeEventType Type { get; set; }

    /// <summary>
    /// Table name.
    /// </summary>
    public string Table { get; set; } = "";

    /// <summary>
    /// Changed row data.
    /// </summary>
    public Dictionary<string, string?>? Row { get; set; }

    /// <summary>
    /// Previous row data (for updates).
    /// </summary>
    public Dictionary<string, string?>? PreviousRow { get; set; }

    /// <summary>
    /// Timestamp of the change.
    /// </summary>
    public DateTime Timestamp { get; set; }

    /// <summary>
    /// Transaction ID if applicable.
    /// </summary>
    public string? TransactionId { get; set; }

    /// <summary>
    /// Position in the change stream.
    /// </summary>
    public long Position { get; set; }
}

/// <summary>
/// Configuration for change streams.
/// </summary>
public class VedaChangeStreamConfig
{
    /// <summary>
    /// Polling interval. Default is 1 second.
    /// </summary>
    public TimeSpan PollInterval { get; set; } = TimeSpan.FromSeconds(1);

    /// <summary>
    /// Maximum events to return per batch. Default is 100.
    /// </summary>
    public int BatchSize { get; set; } = 100;

    /// <summary>
    /// Event types to filter for. Empty means all types.
    /// </summary>
    public List<ChangeEventType>? FilterTypes { get; set; }

    /// <summary>
    /// Start from a specific position. 0 means from the beginning.
    /// </summary>
    public long StartPosition { get; set; }
}

/// <summary>
/// Change stream for watching table changes in real-time.
/// Implements IAsyncEnumerable for streaming consumption.
/// </summary>
public class VedaChangeStream : IAsyncEnumerable<VedaChangeEvent>, IDisposable
{
    private readonly VedaClient _client;
    private readonly string _table;
    private readonly VedaChangeStreamConfig _config;
    private bool _disposed;
    private long _lastPosition;

    /// <summary>
    /// Create a change stream.
    /// </summary>
    public VedaChangeStream(VedaClient client, string table, VedaChangeStreamConfig? config = null)
    {
        _client = client ?? throw new ArgumentNullException(nameof(client));
        _table = table ?? throw new ArgumentNullException(nameof(table));
        _config = config ?? new VedaChangeStreamConfig();
        _lastPosition = _config.StartPosition;
    }

    /// <summary>
    /// Poll for changes once and return events.
    /// </summary>
    public async Task<List<VedaChangeEvent>> PollOnceAsync(CancellationToken ct = default)
    {
        if (_disposed) throw new ObjectDisposedException(GetType().Name);

        var result = await _client.QueryAsync(
            $"WATCH {_table} FROM {_lastPosition} LIMIT {_config.BatchSize}");

        var events = new List<VedaChangeEvent>();
        foreach (var dict in result.ToDicts())
        {
            var evt = ParseChangeEvent(dict);
            if (evt != null)
            {
                events.Add(evt);
                _lastPosition = evt.Position;
            }
        }

        if (events.Count > 0)
        {
            VedaMetrics.Increment("vedadb_changestream_events", events.Count,
                new() { { "table", _table } });
        }

        return events;
    }

    /// <summary>
    /// Get the IAsyncEnumerator for streaming changes.
    /// </summary>
    public IAsyncEnumerator<VedaChangeEvent> GetAsyncEnumerator(CancellationToken ct = default)
    {
        return new ChangeStreamEnumerator(this, ct);
    }

    private VedaChangeEvent? ParseChangeEvent(Dictionary<string, string?> dict)
    {
        try
        {
            var evt = new VedaChangeEvent
            {
                Table = _table,
                Timestamp = dict.TryGetValue("timestamp", out var ts) && DateTime.TryParse(ts, out var dt)
                    ? dt : DateTime.UtcNow,
                TransactionId = dict.GetValueOrDefault("txid")
            };

            if (dict.TryGetValue("type", out var typeStr))
            {
                evt.Type = typeStr?.ToLowerInvariant() switch
                {
                    "insert" => ChangeEventType.Insert,
                    "update" => ChangeEventType.Update,
                    "delete" => ChangeEventType.Delete,
                    "schema" => ChangeEventType.Schema,
                    _ => ChangeEventType.Unknown
                };
            }

            if (dict.TryGetValue("position", out var posStr) && long.TryParse(posStr, out var pos))
                evt.Position = pos;

            if (dict.TryGetValue("row", out var rowJson) && !string.IsNullOrEmpty(rowJson))
            {
                try
                {
                    evt.Row = System.Text.Json.JsonSerializer.Deserialize<Dictionary<string, string?>>(rowJson);
                }
                catch { /* Best effort */ }
            }

            if (dict.TryGetValue("previous", out var prevJson) && !string.IsNullOrEmpty(prevJson))
            {
                try
                {
                    evt.PreviousRow = System.Text.Json.JsonSerializer.Deserialize<Dictionary<string, string?>>(prevJson);
                }
                catch { /* Best effort */ }
            }

            if (_config.FilterTypes != null && _config.FilterTypes.Count > 0 && !_config.FilterTypes.Contains(evt.Type))
                return null;

            return evt;
        }
        catch
        {
            return null;
        }
    }

    private class ChangeStreamEnumerator : IAsyncEnumerator<VedaChangeEvent>
    {
        private readonly VedaChangeStream _stream;
        private readonly CancellationToken _ct;
        private Queue<VedaChangeEvent> _buffer = new();

        public ChangeStreamEnumerator(VedaChangeStream stream, CancellationToken ct)
        {
            _stream = stream;
            _ct = ct;
        }

        public VedaChangeEvent Current { get; private set; } = null!;

        public async ValueTask<bool> MoveNextAsync()
        {
            while (! _ct.IsCancellationRequested)
            {
                if (_buffer.Count > 0)
                {
                    Current = _buffer.Dequeue();
                    return true;
                }

                var events = await _stream.PollOnceAsync(_ct);
                foreach (var evt in events)
                    _buffer.Enqueue(evt);

                if (_buffer.Count == 0)
                {
                    try
                    {
                        await Task.Delay(_stream._config.PollInterval, _ct);
                    }
                    catch (OperationCanceledException) { break; }
                }
            }

            return false;
        }

        public ValueTask DisposeAsync() => ValueTask.CompletedTask;
    }

    public void Dispose()
    {
        if (_disposed) return;
        _disposed = true;
        GC.SuppressFinalize(this);
    }
}
